### PR TITLE
Add serverless example hosted on AWS Lambda

### PR DESCRIPTION
**Pull request by priyankpat**
I was able to get next.js working with serverless as requested by #4482.
This example uses with-react-i18next as a scenario where the express server is not initialized right away.

### CODE REVIEW

**Comment by timneutkens** [comment]
Can we use `hello-world` as a base? 🤔 To not make things overly complex.

**Comment by PR1YANKPAT3L** [comment]
@timneutkens Updated example with simple hello-world.

**Review by timneutkens — CHANGES_REQUESTED** [review]
I guess the i18n code can be removed then

**Review by priyankpat — COMMENTED** [review]
Remove i18n code from the example and updated README instructions.

File: examples/with-aws-serverless/server.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+const express = require('express')
+const next = require('next')
+
+const dev = process.env.NODE_ENV !== 'production'
+const app = next({ dev })
+const handle = app.getRequestHandler()
```
Comment (timneutkens) [inline examples/with-aws-serverless/server.js]: `app.prepare()` is missing here.

File: examples/with-aws-serverless/serverless.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  runtime: nodejs6.10
+  memorySize: 256
+  timeout: 15
+  environment:
+    NODE_ENV: production
+
+functions:
+  nextJsRootApp:
+    handler: lambda.handler
+    events:
```
Comment (Skaronator) [inline examples/with-aws-serverless/serverless.yml]: You can merge `nextJsRootApp` and `nextJsAnyApp` events - just change the events to this:

```diff
    events:
     - http: ANY {proxy+}
+    - http: ANY /
```

**Review by Skaronator — CHANGES_REQUESTED** [review]

File: examples/with-aws-serverless/serverless.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
+service: nextjs-aws-serverless
+
+package:
+  exclude:
```
Comment (Skaronator) [inline examples/with-aws-serverless/serverless.yml]: Please add `yarn.lock` as well.

File: examples/with-aws-serverless/.babelrc [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+    "presets": [
+      "next/babel"
+    ],
+    "plugins": [
+      "transform-decorators-legacy"
+    ]
+  }
```
Comment (Skaronator) [inline examples/with-aws-serverless/.babelrc]: Remove space before `}`

File: examples/with-aws-serverless/serverless.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  serverless-offline:
+    port: 8000
+    dontPrintOutput: true
+  apigwBinary:
+    types:
+      - '*/*'
+
+provider:
+  name: aws
+  runtime: nodejs6.10
```
Comment (Skaronator) [inline examples/with-aws-serverless/serverless.yml]: You should change this to Node 8.10 (better performance & cost less)

**Review by Skaronator — CHANGES_REQUESTED** [review]

**Comment by lolcoolkat** [comment]
I'm assuming we'll be waiting for https://github.com/zeit/next.js/issues/4496 in order for this to work, right?  :D

**Comment by Skaronator** [comment]
It works without #4496 like any other node.js server with next. (Just not as fast as with #4496)

**Comment by lolcoolkat** [comment]
@Skaronator I tried to test this out.. however it didn't work? After I deployed the application and went to the site.. and tried to click the link.. it gave a "Forbidden" message? Any ideas on that?

**Comment by Skaronator** [comment]
Check out this repo: https://github.com/mitchellhuang/future

The files you need should be: lambda.js, next.config.js and server.js + serverless config

**Comment by lolcoolkat** [comment]
@Skaronator Is the domain with the api gateway required? :)

**Comment by lolcoolkat** [comment]
Is the CloudFront distribution created automatically? It outputted a Cloudfront url but didn't actually create it? Please check this http://i.prntscr.com/TXTMfnbIScKTNN8sE9GLJw.png

**Comment by Skaronator** [comment]
CloudFront is not required. (I think it doesn't even work since it is just a CDN)

Looks like you're trying to deploy mitchellhuang/future but I've never done that. Only copyed the relevant parts out of it.

**Comment by lolcoolkat** [comment]
@Skaronator idk still couldn't get it working even after putting your stuff in.. the index page works fine..but after I click a link to go to a new route it just says Forbidden. I guess i'll just wait until an official example is released.

**Comment by skriems** [comment]
I think I've found the culprit here @PR1YANKPAT3L . If you open up your browsers dev tools you'll notice that scripts fail to load from `<url>/_next/...` and the reason for that is, that the `provider.stage` defined in the `serverless.yml` file gets appended to the URL. So basically you need to use an `assetPrefix` here.
Since this is unique to the Lambda environment you can use an env var in your `serverless.yml` and have your `app.setAssetPrefix` in `server.js` based on that env var.

Note that this is not needed if you would use a custom domain via `serverless-custom-domain`. I'm looking into that now.

I've deployed the material-ui nextjs example to Lambda in this repo
https://github.com/skriems/next-material

Best

**Comment by lolcoolkat** [comment]
@skriems Your repo worked great for me! 

**Note**: To get the repo working through Cloudfront and a custom domain, you need to remove the following from the server.js file:

```
if (process.env.LAMBDA) {
  // `assetPrefix` needs to match `serverless.yml:provider.stage
  let assetPrefix = dev ? '/dev' : '/prod'
  app.setAssetPrefix(assetPrefix)
}
```

At least that's what I had to do :)

**Comment by Skaronator** [comment]
Weird... I'm using a custom domain via `serverless-domain-manager` (not a subdomain) and everything just works without modifying the assetPrefix.

**Comment by skriems** [comment]
How odd. I just tried again without adding a prefix and with `basePath: ''` but I get those 403's. Is it related to the region and/or `endpointType` ?

**Comment by Skaronator** [comment]
Ah the  `execute-api.eu-central-1.amazonaws.com` endpoint has a `/prod` which mess up the whole path while the custom domain doesn't have it. Thanks to this mapping:
image

**Comment by skriems** [comment]
exactly that's the problem. Any opinion on how this should be included/documented in the example?

**Comment by Skaronator** [comment]
It might be better to use [Lambda@Edge](https://aws.amazon.com/de/lambda/edge/) instead of Lambda with APIGateway but Serverless doesn't support it yet due to a bug in CloudFormation.

https://github.com/serverless/serverless/pull/4796

**Comment by Enalmada** [comment]
@skriems [next-material](https://github.com/skriems/next-material) serverless example worked great.  If the requested changes in this pull are not being acted on promptly, can we just get that into the next.js examples section?      

Also, I see a few serverless example request tickets imply that there might be good reason for a next.js example to handle the "scenario where the express server is not initialized right away."   Can any of the serverless hackers in this thread confirm if a serverless example based on "with-react-i18next" would be different in any fundamental way than one based on "hello-world"?  Does the [next-material](https://github.com/skriems/next-material) example use an express server not initialized right away?

**Comment by romainquellec** [comment]
> between 150ms and 2.5s when "warm".

If you have 2.5s of response time, you did something wrong with your lambda. Waiting for https://github.com/zeit/next.js/issues/4496 though !

**Comment by romainquellec** [comment]
also, why serverless-http over aws-serverless-express ?

**Comment by Enalmada** [comment]
> If you have 2.5s of response time, you did something wrong with your lambda. 

@romainquellec It is certainly possible and testing was extremely unscientific.  So everyone can do their own testing I will elaborate slightly:  If I refresh my serverless next.js example quickly (once a second) the responses definitely tend to average closer to 150ms.   If I wait a few minutes between each refresh (small enough interval that containers should be warm, trying to replicate my actual site traffic pattern), responses vary wildly between 150ms and 2.5s.  YMMV and this is just food for thought as I was just messing around.

**Comment by timneutkens** [comment]
I'm going to close this PR in favor of the work we're doing towards making Next.js output single-route lambdas.

**Pull request closed without merging**